sklearn/metrics/cluster/tests/test_unsupervised.py

```diff
@@ -15,31 +16,43 @@
 def test_silhouette():
     # Tests the Silhouette Coefficient.
     dataset = datasets.load_iris()
-    X = dataset.data
+    X_dense = dataset.data
+    X_csr = csr_matrix(X_dense)
+    X_dok = sp.dok_matrix(X_dense)
+    X_lil = sp.lil_matrix(X_dense)
     y = dataset.target
-    D = pairwise_distances(X, metric='euclidean')
-    # Given that the actual labels are used, we can assume that S would be
-    # positive.
-    silhouette = silhouette_score(D, y, metric='precomputed')
-    assert(silhouette > 0)
-    # Test without calculating D
-    silhouette_metric = silhouette_score(X, y, metric='euclidean')
-    assert_almost_equal(silhouette, silhouette_metric)
-    # Test with sampling
-    silhouette = silhouette_score(D, y, metric='precomputed',
-                                  sample_size=int(X.shape[0] / 2),
-                                  random_state=0)
-    silhouette_metric = silhouette_score(X, y, metric='euclidean',
-                                         sample_size=int(X.shape[0] / 2),
-                                         random_state=0)
-    assert(silhouette > 0)
-    assert(silhouette_metric > 0)
-    assert_almost_equal(silhouette_metric, silhouette)
-    # Test with sparse X
-    X_sparse = csr_matrix(X)
-    D = pairwise_distances(X_sparse, metric='euclidean')
-    silhouette = silhouette_score(D, y, metric='precomputed')
-    assert(silhouette > 0)
+
+    for X in [X_dense, X_csr, X_dok, X_lil]:
+        D = pairwise_distances(X, metric='euclidean')
+        # Given that the actual labels are used, we can assume that S would be
+        # positive.
+        score_precomputed = silhouette_score(D, y, metric='precomputed')
+        assert(score_precomputed > 0)
+        # Test without calculating D
+        score_euclidean = silhouette_score(X, y, metric='euclidean')
+        assert_almost_equal(score_precomputed, score_euclidean)
+
+        if X is X_dense:
+            score_dense_without_sampling = score_precomputed
+        else:
+            assert_almost_equal(score_euclidean,
+                                score_dense_without_sampling)
+
+        # Test with sampling
+        score_precomputed = silhouette_score(D, y, metric='precomputed',
+                                             sample_size=int(X.shape[0] / 2),
+                                             random_state=0)
+        score_euclidean = silhouette_score(X, y, metric='euclidean',
+                                           sample_size=int(X.shape[0] / 2),
+                                           random_state=0)
+        assert(score_precomputed > 0)
+        assert(score_euclidean > 0)
```

Comment:
This PR is a great opportunity to use `assert_greater` to check those assertions and get meaningful error messages in case of future regression.